pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go

```diff
@@ -147,23 +150,49 @@ func errMsgPrefix(b BuildCtx, id catid.DescID) string {
 
 // dropCascadeDescriptor contains the common logic for dropping something with
 // CASCADE.
-func dropCascadeDescriptor(b BuildCtx, id catid.DescID) {
+func dropCascadeDescriptor(b BuildCtx, n tree.NodeFormatter, id catid.DescID) {
+	var s dropCascadeState
+	// First, we recursively visit the descriptors to which the drop cascades.
+	s.visitRecursive(b, id)
+	// Do all known whole-descriptor drops first.
+	// This way, we straightforwardly set a large number of to-ABSENT targets.
+	for _, qe := range s.q {
+		qe.dropWholeDescriptor()
+	}
+	// Afterwards, we can deal with any remaining column, index, constraint or
+	// other element removals when their parent descriptor-element isn't dropped.
+	for _, qe := range s.q {
+		qe.dropUndroppedBackReferencedElements(n)
+	}
+}
```

Comment:
I had to rewrite this to drop the descriptors wholesale in a first pass before fiddling with individual column drops which may otherwise not work because zone configs or whatever. This change is worth keeping as it's more sensible anyways.